src/documents/tasks.py

```diff
@@ -515,3 +515,41 @@ def check_scheduled_workflows():
                             workflow_to_run=workflow,
                             document=document,
                         )
+
+
+def update_document_parent_tags(tag: Tag, new_parent: Tag) -> None:
+    """
+    When a tag's parent changes, ensure all documents containing the tag also have
+    the parent tag (and its ancestors) applied.
+    """
+    doc_tag_relationship = Document.tags.through
+
+    doc_ids: list[int] = list(
+        Document.objects.filter(tags=tag).values_list("pk", flat=True),
+    )
+
+    if not doc_ids:
+        return
+
+    parents_to_add = [new_parent, *new_parent.get_ancestors()]
+
+    to_create: list = []
+    affected: set[int] = set()
+
+    for parent in parents_to_add:
```

Comment:
Maybe reverse this thinking?  Get the existing relationships, then create only the missing ones, instead of checking each for the missing.

```python
parent_ids = [new_parent.id] + new_parent.get_ancestors_pks()

existing_pairs = set(
    doc_tag_relationship.objects
    .filter(document_id__in=doc_ids, tag_id__in=parent_ids)
    .values_list("document_id", "tag_id")
)
for doc_id in doc_ids:
    for parent_id in parent_ids:
        if (doc_id, parent_id) not in existing_pairs:
            to_create.append(
                doc_tag_relationship(document_id=doc_id, tag_id=parent_id)
            )
            affected.add(doc_id)
```

Something like this should reduce the queries again, so we get existing in 1, then bulk create in another.